File: Python/setZeroes.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: setZeroes.py
@Function: 矩阵置零
@Link: https://leetcode-cn.com/problems/set-matrix-zeroes/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-30
"""

from typing import List


class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        m, n = len(matrix), len(matrix[0])
        flag_col0 = False

        for i in range(m):
            if matrix[i][0] == 0:
                flag_col0 = True
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = matrix[0][j] = 0

        for i in range(m - 1, -1, -1):
            for j in range(1, n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0
            if flag_col0:
                matrix[i][0] = 0


# -------------------------
if __name__ == "__main__":
    matrix = [[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]]

    solution = Solution()
    print(f"Processing before: \n{matrix}")
    solution.setZeroes(matrix)
    print(f"Processing after: \n{matrix}")

File: C++/setZeroes.cpp
/**
 * @File    : setZeroes.cpp
 * @Brief   : 矩阵置零
 * @Link    : https://leetcode-cn.com/problems/set-matrix-zeroes/
 * @Author  : Wei Li
 * @Date    : 2021-08-30
*/

#include <iostream>
#include <vector>

class Solution
{
public:
    void setZeroes(std::vector<std::vector<int>> &matrix)
    {
        int m = matrix.size();
        int n = matrix[0].size();
        int flag_col0 = false;
        for (int i = 0; i < m; i++)
        {
            if (!matrix[i][0])
            {
                flag_col0 = true;
            }
            for (int j = 1; j < n; j++)
            {
                if (!matrix[i][j])
                {
                    matrix[i][0] = matrix[0][j] = 0;
                }
            }
        }
        for (int i = m - 1; i >= 0; i--)
        {
            for (int j = 1; j < n; j++)
            {
                if (!matrix[i][0] || !matrix[0][j])
                {
                    matrix[i][j] = 0;
                }
            }
            if (flag_col0)
            {
                matrix[i][0] = 0;
            }
        }
    }
};

template <typename T>
void DisplayElement(const T &container)
{
    for (auto element : container)
    {
        for (auto value : element)
        {
            std::cout << value << " ";
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

// ----------------------------
int main(int argc, char **argv)
{
    std::vector<std::vector<int>> matrx = {{1, 1, 1}, {1, 0, 1}, {1, 1, 1}};

    Solution solution;
    std::cout << "Processing before : " << std::endl; 
    DisplayElement(matrx);
    solution.setZeroes(matrx);
    std::cout << "Processing after : " << std::endl; 
    DisplayElement(matrx);

    return 0;
}
